internal/monitors/cpu/cpu.go

```diff
@@ -0,0 +1,239 @@
+		return
+	}
+
+	usedDiff := current.Used - prev.Used
+	totalDiff := current.Total - prev.Total
+	if usedDiff < 0 || totalDiff < 0 {
+		err = fmt.Errorf("usedDiff (%v) or totalDiff (%v) are < 0", usedDiff, totalDiff)
+	} else if (usedDiff == 0 && totalDiff == 0) || totalDiff == 0 {
+		utilization = 0
+	} else {
+		// calculate utilization
+		utilization = usedDiff / totalDiff * 100
+		if utilization < 0 {
+			err = fmt.Errorf("percent %v < 0 total: %v used: %v", utilization, usedDiff, totalDiff)
+		}
+		if utilization > 100 {
+			err = fmt.Errorf("percent %v > 100 total: %v used: %v ", utilization, usedDiff, totalDiff)
+		}
+	}
+
+	return
+}
+
+func (m *Monitor) initializeCPUTimes() {
+	// initialize previous values
+	var total []cpu.TimesStat
+	var err error
+	if total, err = times(false); err != nil {
+		logger.WithField("warning", err).Warningf("unable to initialize cpu times will try again in the next reporting cycle")
+	}
+	if len(total) > 0 {
+		m.previousTotal = cpuTimeStatTototalUsed(&total[0])
+	}
+}
+
+func (m *Monitor) initializePerCoreCPUTimes() {
+	// initialize per core cpu times
+	var totals []cpu.TimesStat
+	var err error
+	if totals, err = times(true); err != nil {
+		logger.WithField("warning", err).Warningf("unable to initialize per core cpu times will try again in the next reporting cycle")
+	}
+	m.previousPerCore = make(map[string]*totalUsed, len(totals))
+	for _, core := range totals {
+		m.previousPerCore[core.CPU] = cpuTimeStatTototalUsed(&core)
+	}
+}
+
+// Configure is the main function of the monitor, it will report host metadata
+// on a varied interval
+func (m *Monitor) Configure(conf *Config) error {
+	if runtime.GOOS != "windows" {
+		logger.Warningf("'%s' monitor is in beta on this platform.  For production environments please use 'collectd/%s'.", monitorType, monitorType)
+	}
+
+	// create contexts for managing the the plugin loop
+	var ctx context.Context
+	ctx, m.cancel = context.WithCancel(context.Background())
+
+	// save config to monitor for convenience
+	m.conf = conf
+
+	// initialize cpu times and per core cpu times so that we don't have to wait an entire reporting interval to report utilization
+	m.initializeCPUTimes()
+	m.initializePerCoreCPUTimes()
+
+	// gather metrics on the specified interval
+	utils.RunOnInterval(ctx, func() {
+		m.emitDatapoints()
+		// NOTE: If this monitor ever fails to complete in a reporting interval
+		// maybe run this on a separate go routine
+		m.emitPerCoreDatapoints()
+	}, time.Duration(conf.IntervalSeconds)*time.Second)
+
+	return nil
+}
+
+// Shutdown stops the metric sync
+func (m *Monitor) Shutdown() {
+	if m.cancel != nil {
+		m.cancel()
+	}
+}
+
+// cpuTimeStatTototalUsed converts a cpu.TimesStat to a totalUsed with Total and Used values
+func cpuTimeStatTototalUsed(t *cpu.TimesStat) *totalUsed {
+	// add up all times if a value doesn't apply then the struct field
+	// will be 0 and shouldn't affect anything
+	total := t.User +
+		t.System +
+		t.Idle +
```

Comment:
Should idle time really be added to the "total' value?

Reply:
yes because it's part of the total.  We then get the usage by subtracting the idle from the total.